### PR TITLE
Require Query IDs over inline query JSON

### PR DESCRIPTION
**Pull request by jharley**
As discussed in #96, Triggers and Boards now require Query IDs instead of JSON.

### CODE REVIEW

**Comment by jharley** [comment]
Requires #98

**Review by dstrelau — APPROVED** [review]
Nothing blocking. Some doc example fixes and a few questions.

File: docs/resources/board.md [hunk 1]
```diff
@@ -14,7 +14,7 @@ locals {
 }
 
 data "honeycombio_query_specification" "query" {
-  count = length(local.percentiles)
+  for_each = local.percentiles
 
   calculation {
     op     = local.percentiles[count.index]
```
[the remark this answers is not in the document]
Comment (jharley) [reply docs/resources/board.md:20]: I love that idea, opened https://github.com/honeycombio/terraform-provider-honeycombio/issues/104

File: docs/resources/board.md [hunk 2]
```diff
@@ -14,7 +14,7 @@ locals {
 }
 
 data "honeycombio_query_specification" "query" {
-  count = length(local.percentiles)
+  for_each = local.percentiles
```
[the remark this answers is not in the document]
Comment (jharley) [reply docs/resources/board.md]: Good catches, thank you!

File: client/trigger.go [hunk 3]
```diff
@@ -144,6 +146,30 @@ func TriggerRecipientTypes() []TriggerRecipientType {
...
 }
 
+func (t *Trigger) MarshalJSON() ([]byte, error) {
+	// aliased type to avoid stack overflows due to recursion
+	type ATrigger Trigger
+
+	if t.QueryID != "" && t.Query != nil {
+		// we can't sent both to the API, so favour QueryID
+		// this doesn't work in the general case, but this
+		// client is now purpose-built for the Terraform provider
```
[the remark this answers is not in the document]
Comment (jharley) [reply client/trigger.go:156]: API funkiness: the general recommended pattern writing providers is to create or update a resource and then read it back putting the read contents into the state. The Trigger and Board APIs both return both query id and the inline query regardless of if it was created with inline JSON or an ID:

```sh
$ curl -s https://api.honeycomb.io/1/triggers/testacc/h6hZcHW91jd -H "X-Honeycomb-Team: $HONEYCOMBIO_APIKEY" | jq -r
{
  "name": "Test trigger from terraform-provider-honeycombio",
  "frequency": 300,
  "disabled": false,
  "query": {
    "calculations": [
      {
        "op": "COUNT"
      }
    ],
    "time_range": 1200
  },
  "query_id": "qjMXte24eSx",
  "threshold": {
    "op": ">",
    "value": 100
  },
  "id": "h6hZcHW91jd",
  "triggered": false,
  "AllowRateTriggers": false
}```

File: honeycombio/resource_trigger.go [hunk 4]
```diff
@@ -40,16 +39,9 @@ func newTrigger() *schema.Resource {
...
+			"query_id": {
 				Type:     schema.TypeString,
 				Required: true,
-				ValidateDiagFunc: validateQueryJSON(func(q *honeycombio.QuerySpec) diag.Diagnostics {
-					err := honeycombio.MatchesTriggerSubset(q)
-					if err != nil {
-						return diag.FromErr(err)
-					}
-					return nil
-				}),
```
[the remark this answers is not in the document]
Comment (jharley) [reply honeycombio/resource_trigger.go:52]: We answered this via Slack, but closing the loop: we confirmed this is a gap at the moment.

**Pull request merged**